src/Microsoft.CodeAnalysis.Razor/DefaultTagHelperDescriptorFactory.cs

```diff
@@ -61,87 +61,44 @@ public DefaultTagHelperDescriptorFactory(Compilation compilation, bool designTim
+                return null;
             }
 
-            var attributeDescriptors = GetAttributeDescriptors(type, errorSink);
-            var targetElementAttributes = GetValidHtmlTargetElementAttributes(type, errorSink);
-            var allowedChildren = GetAllowedChildren(type, errorSink);
+            var typeName = GetFullName(type);
+            var assemblyName = type.ContainingAssembly.Identity.Name;
+            var descriptorBuilder = ITagHelperDescriptorBuilder.Create(typeName, assemblyName);
 
-            var tagHelperDescriptors =
-                BuildTagHelperDescriptors(
-                    type,
-                    type.ContainingAssembly.Identity.Name,
-                    attributeDescriptors,
-                    targetElementAttributes,
-                    allowedChildren);
+            AddBoundAttributes(type, descriptorBuilder);
+            AddTagMatchingRules(type, descriptorBuilder);
+            AddAllowedChildren(type, descriptorBuilder);
+            AddDocumentation(type, descriptorBuilder);
+            AddTagOutputHint(type, descriptorBuilder);
 
-            return tagHelperDescriptors.Distinct(TagHelperDescriptorComparer.Default);
-        }
-
-        private IEnumerable<AttributeData> GetValidHtmlTargetElementAttributes(
-            INamedTypeSymbol typeSymbol,
-            ErrorSink errorSink)
-        {
-            var targetElementAttributes = typeSymbol.GetAttributes().Where(a => a.AttributeClass == _htmlTargetElementAttributeSymbol);
-            return targetElementAttributes.Where(a => ValidHtmlTargetElementAttributeNames(a, errorSink));
-        }
-
-        private IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
-            INamedTypeSymbol type,
-            string assemblyName,
-            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
-            IEnumerable<AttributeData> targetElementAttributes,
-            IEnumerable<string> allowedChildren)
-        {
-            TagHelperDesignTimeDescriptor designTimeDescriptor = null;
-            if (DesignTime)
+            var diagnostics = descriptorBuilder.Validate();
+            foreach (var diagnostic in diagnostics)
             {
-                XmlMemberDocumentation documentation = null;
-                var xml = type.GetDocumentationCommentXml();
-                if (!string.IsNullOrEmpty(xml))
-                {
-                    documentation = new XmlMemberDocumentation(xml);
-                }
-
-                string outputElementHint = null;
-                var outputElementHintAttribute = type.GetAttributes().Where(a => a.AttributeClass == _outputElementHintAttributeSymbol).FirstOrDefault();
-                if (outputElementHintAttribute != null)
-                {
-                    outputElementHint = (string)(outputElementHintAttribute.ConstructorArguments[0]).Value;
-                }
+                descriptorBuilder.AddDiagnostic(diagnostic);
+            }
```

Comment:
I commented about this elsewhere, but this shouldn't be necessary (L86-L90). This should just be part of build.